// css-entry-webpack-plugin/lib/interop/webpack/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.webpackDevServerResourceTests = exports.multiEntryDependencyLocSeparator = exports.MultiEntryDependency = exports.SingleEntryDependency = exports.Chunk = exports.MultiModule = exports.Module = exports.Compilation = exports.Compiler = exports.Source = exports.Resolver = undefined;
exports.loaderToIdent = loaderToIdent;
exports.loadersToRequestIdent = loadersToRequestIdent;
exports.isWebpackDevServerResource = isWebpackDevServerResource;
exports.excludeWebpackDevServerResources = excludeWebpackDevServerResources;

var _webpack = require("webpack");

var _Compiler = require("webpack/lib/Compiler");

var _Compiler2 = _interopRequireDefault(_Compiler);

var _Compilation = require("webpack/lib/Compilation");

var _Compilation2 = _interopRequireDefault(_Compilation);

var _Module = require("webpack/lib/Module");

var _Module2 = _interopRequireDefault(_Module);

var _MultiModule = require("webpack/lib/MultiModule");

var _MultiModule2 = _interopRequireDefault(_MultiModule);

var _Chunk = require("webpack/lib/Chunk");

var _Chunk2 = _interopRequireDefault(_Chunk);

var _SingleEntryDependency = require("webpack/lib/dependencies/SingleEntryDependency");

var _SingleEntryDependency2 = _interopRequireDefault(_SingleEntryDependency);

var _MultiEntryDependency = require("webpack/lib/dependencies/MultiEntryDependency");

var _MultiEntryDependency2 = _interopRequireDefault(_MultiEntryDependency);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Resolver = _webpack.Resolver;
exports.Source = _webpack.Source;
exports.Compiler = _Compiler2.default;
exports.Compilation = _Compilation2.default;
exports.Module = _Module2.default;
exports.MultiModule = _MultiModule2.default;
exports.Chunk = _Chunk2.default;
exports.SingleEntryDependency = _SingleEntryDependency2.default;
exports.MultiEntryDependency = _MultiEntryDependency2.default;
const multiEntryDependencyLocSeparator = exports.multiEntryDependencyLocSeparator = ":";
const webpackDevServerResourceTests = exports.webpackDevServerResourceTests = [/webpack-dev-server(\/|\\)client/, /webpack\/hot/];
function loaderToIdent(data) {
    if (!data.options) return data.loader;
    if (typeof data.options === "string") return data.loader + "?" + data.options;
    if (typeof data.options !== "object") throw new Error("loader options must be string or object");
    if (data.ident) return data.loader + "??" + data.ident;
    return data.loader + "?" + JSON.stringify(data.options);
}
function loadersToRequestIdent(loaders, resource) {
    return loaders.map(loaderToIdent).concat([resource]).join("!");
}
function isWebpackDevServerResource(resource) {
    return webpackDevServerResourceTests.some(resourceTest => resourceTest.test(resource));
}
function excludeWebpackDevServerResources(resources) {
    return resources.filter(resource => !isWebpackDevServerResource(resource));
}
